### PR TITLE
feat: Add compatibility for MongoDB Atlas Serverless and AWS Amazon DocumentDB with collation options `enableCollationCaseComparison`, `convertEmailToLowercase`, `convertUsernameToLowercase`

### LINKED ISSUE #7937: Parser Server not fully compatible with Mongo Atlas Serverless
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description
I currently checking if parse-server could run in a fully serverless env.
The experiment is based on GCP Cloud Run and the new Mongo Atlas Serverless service (https://www.mongodb.com/cloud/atlas/serverless)

I found an issue (no current workarounds) about collation (case insensitive login/query/indexes). Mongo Serverless does not support collation. Developers are not able to "disable" collation. The collation fails will for example block logins, and block Defined Schema migration.

### Steps to reproduce
Run a parse-server and connect to a Mongo Serverless Instance

### Actual Outcome
All stuff related to case insensitive query/index will fail

### Expected Outcome
To be able to disable the collation (mainly used on the Parse. User class)
### Environment
[5.3.0-alpha.6](https://www.npmjs.com/package/parse-server/v/5.3.0-alpha.6)

Server
- Parse Server version: `5.3.0-alpha.6`

Database
- System (MongoDB or Postgres): `Mongo Atlas Serverless`
- Database version: `5.3.1`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `remote`

### Logs
Collation is not supported on this Altas tier (Mongo Serverless)

### PR DESCRIPTION
**Pull request by mattia1208**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

## Issue
Add options to support MongoDB Serverless and AWS DocumentDB which currently does not have collation support.

Closes: #7937 

## Approach
Add 3 abstract options disableCollation, transformEmailToLowerCase and transformUsernameToLowerCase to allow a developer to disable default collations of Parse Server and transformEmailToLowerCase to keep clean emails in the database and transformUsernameToLowerCase to keep clean usernames in the database.

## Tasks

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, code comments)
- [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)

### CODE REVIEW

**Comment by mattia1208** [comment]
This PR is based on:

https://github.com/parse-community/parse-server/pull/8042
Original author: @Moumouls

And 

https://github.com/parse-community/parse-server/pull/8767
Original author: @b10f

File: spec/DatabaseController.spec.js
Comment (Sbef98) [inline spec/DatabaseController.spec.js:366]: Maybe it needs to be written calling `function()` instead of the short version like stated by @mtrezza to b10f?

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* Optional. If set to `true`, the `email` property of a user is automatically converted to lowercase before being stored in the database. Consequently, queries must match the case as stored in the database, which would be lowercase in this scenario. If `false`, the `email` property is stored as set, without any case modifications. Default is `false`.
  :DEFAULT: false */
  convertEmailToLowercase: ?boolean;
   /* Optional. If set to `true`, the `username` property of a user is automatically converted to lowercase before being stored in the database. Consequently, queries must match the case as stored in the database, which would be lowercase in this scenario. If `false`, the `username` property is stored as set, without any case modifications. Default is `false`.
  :DEFAULT: false */
  convertUsernameToLowercase: ?boolean;
```

- Extended the description.
- Added missing `:DEFAULT:` values; these guarantee that also in code there is always a default value and the properties are never `undefined` - consider in code and tests if necessary.
- A simply change in case is more commonly referred to as a "conversion" in technical literature.
- Removed uppercase in var name `...ToLowerCase`, as "lowercase" is 1 word.

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* Optional. If set to `true`, the collation rule of case comparison for queries and indexes is enabled. Enable this option to run Parse Server with MongoDB Atlas Serverless or AWS Amazon DocumentDB. If `false`, the collation rule of case comparison is disabled. Default is `false`.
  :DEFAULT: false */
  enableCollationCaseComparison: ?boolean;
```

- The term "disable collation" does not have a clear meaning. A "collation" is a set of rules for how to compare strings. Therefore, collation is never disabled, its just set to be case sensitive or insensitive. Also, to "disable case insensitivity" is a double negation, which makes it more difficult to understand than simply "enable case comparison".
- renamed to `enableCollationCaseComparison` so we can easier group future collation related options `enableCollation...`.
- fixed product marketing names

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by mtrezza** [comment]
@mattia1208 Thanks for picking up this PR; just a few cosmetic things and this should be good to merge.

**Review by mtrezza — CHANGES_REQUESTED** [review]
Could you run `npm run definitions` to update the definition files? There also seem to be some lint errors where the new option names are not changed in code.

**Comment by mattia1208** [comment]
@mtrezza sorry, done!

**Review by mtrezza — DISMISSED** [review]
Looks good; waiting for CI to pass...

**Comment by mtrezza** [comment]
CI fails with:

```
1) DatabaseController convertEmailToLowercase should not find a case insensitive user by email with convertEmailToLowercase
  - Error: bad or missing username

2) DatabaseController enableCollationCaseComparison should not create insensitive indexes with enableCollationCaseComparison
  - Expected object not to have properties
      case_insensitive_username: Object({ username: 1 })
      case_insensitive_email: Object({ email: 1 })

3) DatabaseController enableCollationCaseComparison should force caseInsensitive to false with enableCollationCaseComparison option
  - Expected true to equal false.
```

Did the tests pass locally for you?

**Comment by mattia1208** [comment]
@mtrezza now all tests locally have passed

**Comment by mtrezza** [comment]
Restarted CI...

**Comment by mtrezza** [comment]
CI passed; @Moumouls, @b10f do you want to review this once more before we merge?

**Comment by b10f** [comment]
> CI passed; @Moumouls, @b10f do you want to review this once more before we merge?

I'm good

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.4.0-alpha.8](https://github.com/parse-community/parse-server/releases/tag/6.4.0-alpha.8)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.5.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.5.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.5.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/6.5.0-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.5.0](https://github.com/parse-community/parse-server/releases/tag/6.5.0)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.0.0](https://github.com/parse-community/parse-server/releases/tag/7.0.0)

**Comment by Moumouls** [comment]
Ooohhh nice I didn't see this notification, thanks @mattia1208 for your work here ! 

Mongo serverless is really nice and can help a lot to adopt parse server, with super cheap deployment cost.

**Pull request merged**
